### DISCUSSION TITLE
Remove processed background task objects

### DISCUSSION THREAD

**Discussion by jainal09** [encode/starlette#2546] [Potential Issue]
This discussion is based on #2542 

# Summary
When a background task is processed it is not removed from the array and it is very difficult to implement features in which we want to find length of queued background tasks

This can be used to freeze shutdowns when handling OS signals for enabling graceful shutdowns. Thus, if there are already enqueued background tasks wait and only then stop the server once all the tasks are finished.

The reason, for this is because once the task is processed in await task(), why is there a need to keep it in the list. I am removing it because, I want to achieve something like this

If I want to fetch the current task queue length

using something like below

```python

class QueueOut(BaseModel):
    task_lenth: int

@app.get("/get_queue_length/", response_model=QueueOut)
async def read_task():
    return QueueOut(task_lenth=len(tasks.tasks))
```    
This is helpful for graceful shutdown of the application in scenarios where a lots of bg task are enqueued and now you intercepted a SIGTERM or SIGKILL which sends a termination signal but, the server will not be stopped until all the enqueued tasks are processed.

So, I propose to change the `starlette/background.py` 

Line 43
```python
    async def __call__(self) -> None:
        for task in self.tasks.copy():
            await task()
            if task in self.tasks:
                self.tasks.remove(task)
```

**Comment by jainal09** [comment]
@Kludex

**Reply by jainal09** [reply]
@adriangb

**Comment by adriangb** [comment]
I don’t see a problem with the change. I don’t think you need the `if task in self.tasks`. But other than that, it’s fine.

But you could also use your own background task system. Here’s an example https://github.com/adriangb/asgi-background

**Comment by jainal09** [comment]
@adriangb I know it's been more than a year and I apologize for the delay, but I've finally opened a PR for this: https://github.com/Kludex/starlette/pull/3089

Would appreciate your review when you get a chance!

**Reply by Kludex** [reply]
> Thus, if there are already enqueued background tasks wait and only then stop the server once all the tasks are finished.

This already happens.
